### ISSUE TITLE
Unity iOS Mobile Analytics No Attributes / Metrics

### ISSUE THREAD

**Issue by Deathwing** [aws/aws-sdk-net#328]
Hey guys,

we have implemented mobile analytics in our unity3d application.
The tracking on Android is working fine, all attributes and metrics can be found in the developer console and on redshift.
But every event on iOS only consists of the 'custom event name', no metric or attributes are included inside them.

Is this an bug or intended behaviour?

Cheers!

**Comment by johnnyhoffman** [comment]
Hi,
No, this isn't intended behavior. I just double checked with the sample app to ensure that it is working as desired, and all of the custom attributes and metrics are available to me on the console for both Android and iOS. 

Is there any other information you can provide that would help me replicate this behavior? e.g.: Any major difference between the way you use Mobile Analytics compared to the way the sample uses it? Any log data related to Mobile Analytics?

Thanks.

**Comment by phodges** [comment]
Hello,

We're using the 3.1.73.0 SDK and also have this problem: android works fine, but I see nothing other than the plain event reported for iOS.  I would say our use case is typical: dispatching events whenever something of significance has occurred. 

I've considered a couple of possibilities:

- NSAppTransportSecurity settings, which seem fine.
- Ensuring 'allowUseDataNetwork' is set true in awsconfig.xml, since any restriction in the related policy would block delivery (although that would surely have killed all events). 

I've tried watching verbose logs in Xcode and mostly seen expected logs. Typical messages are as follows:

Mobile Analytics client successfully delivered 7 events to service. Delete those events from local storage.
Amazon.Runtime.Internal.Util.Logger:InfoFormat(String, Object[])

Mobile Analytics client successfully delivered 7 events to service. Delete those events from local storage.
Amazon.Runtime.Internal.Util.Logger:InfoFormat(String, Object[])

After a number of events were apparently successfully sent, the following logs appeared:

Mobile Analytics Manager is trying to deliver events in background thread.
Amazon.Runtime.Internal.Util.Logger:InfoFormat(String, Object[])
Amazon.MobileAnalytics.MobileAnalyticsManager.Internal.BackgroundRunner:DoWork()

Policy restriction: ConnectivityPolicy
Amazon.Runtime.Internal.Util.Logger:InfoFormat(String, Object[])
Amazon.MobileAnalytics.MobileAnalyticsManager.Internal.DeliveryClient:AttemptDelivery()
Amazon.MobileAnalytics.MobileAnalyticsManager.Internal.BackgroundRunner:DoWork()
 
The object that coordinates work, sending events to the analytics manager, is a singleton and is never destroyed.

Is there any other information that might be useful to you in tracking down this problem?

**Comment by phodges** [comment]
Today, I took the sample code and inserted my own app and user pool IDs into it. After firing test events from this sample, running on iOS, I found the same results.

**Comment by phodges** [comment]
I narrowed the problem down to a particular position. In DeliveryClient.AttemptDelivery, at the point at which JsonMapper.ToObject<A.MA.M.Event> is invoked, I see valid json passed in and missing attributes and metrics on the deserialised event object.

I will continue to investigate, but I believe this to be an il2cpp bug.

**Comment by phodges** [comment]
I can confirm that this is an il2cpp bug and is reproducible in Unity 5.3.5 f1 and p2. The problem here is that the PropertyInfo retrieved for types is incorrect, in particular 'CanWrite' is returned as false This means JsonMapper.ReadValue will behave unexpectedly (see line 443).

I have reported this problem to Unity, and you may wish to add your voice to the discussion. For reference, here's a link to the bug report: [803913_51oln3u5te2113t7](https://fogbugz.unity3d.com/default.asp?803913_51oln3u5te2113t7). I can also provide you with test case, should you find that useful.

**Comment by phodges** [comment]
Unity have investigated, and this was my own fault and relates to code stripping options. For the benefit of anybody else that runs into the same problem, I had neglected to protect MobileAnalytics from stripping under il2cpp rules. The following line should appear in your link.xml script:

`<assembly fullname="AWSSDK.MobileAnalytics" preserve="all"/>`

This is mentioned in the sample readme.md, so really should not have escaped my attention. Once this line is in place, everything works.

**Comment by karthiksaligrama** [comment]
awesome. Closing this issue for now. Please create a new issue if you face the problem again.
